MotionMark Images subtest rendering is incorrect on ASi (no dissolve effect)

kkinnunen-apple opened the pull request:
#### a4350b5e3b698b76790a7e2c9cc0f58f5ee66fcc
<pre>
MotionMark Images subtest rendering is incorrect on ASi (no dissolve effect)
<a href="https://bugs.webkit.org/show_bug.cgi?id=252482">https://bugs.webkit.org/show_bug.cgi?id=252482</a>
rdar://problem/105606592

Reviewed by Simon Fraser.

Fix rendering of accelerated canvas elements into small layer tiles
after putImageData.

ImageBufferIOSurfaceBackend::putPixelBuffer would clean the CGImage
cache, if there&apos;s possibility that the cache exists.

ImageBufferIOSurfaceBackend::copyNativeImageForDrawing() does not
currently mark the cache as being active. Instead, it relies on
after-the-fact cleanup via
ImageBufferIOSurfaceBackend::finalizeDrawIntoContext().

ImageBufferIOSurfaceBackend::finalizeDrawIntoContext() would only
cleanup if the draw happened on an non-accelerated context.

Accelerated context status would be determined by whether WebKit
requested the layer to be accelerated.

In case the layer is small, CoreAnimation might give a bitmap-backed
CGContext for the drawing context, even though WebKit
requested an accelerated context. WebKit would
mark the context as accelerated, even when it was not.

This would lead into scenario where the CGImage cache would not be
cleaned after each draw, and hence putImageData would not be visible.

Fix by removing the GraphicsContextCG::setIsAcceleratedContext().
Instead, query the actual state from the CGContext.

Adds a test in ImageBufferTests that tests the code paths. This test
would not have caught the issue, since the bug as incorrect use of
setIsAcceleratedContext() rooted from PlatformCALayer::drawLayerContents.
Nevertheless, the test would have given a bit more confidence that the
complicated invalidation logic was correct in the specific case.

Adds a trivial test for the acceleration status query based on the
CGContext.

* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::setIsAcceleratedContext): Deleted.
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayer::drawLayerContents):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::renderingMode const):
(WebCore::GraphicsContextCG::setIsAcceleratedContext): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextPlatformPrivateCG.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::ensureGraphicsContext):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/mac/WebLayer.mm:
(-[WebSimpleLayer drawInContext:]):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _drawRect:contentsOnly:]):
* Tools/TestWebKitAPI/Tests/WebCore/ImageBufferTests.cpp:
(TestWebKitAPI::imageBufferPixelIs):
(TestWebKitAPI::hasTestPattern):
(TestWebKitAPI::drawTestPattern):
(TestWebKitAPI::createPixelBufferTestPattern):
(TestWebKitAPI::toImageBufferOptions):
(TestWebKitAPI::PreserveResolutionOperationTest::imageBufferOptions const):
(TestWebKitAPI::TEST_P):
(TestWebKitAPI::AnyScaleTest::imageBufferOptions const):
(TestWebKitAPI::AnyTwoImageBufferOptionsTest::imageBufferOptions0 const):
(TestWebKitAPI::AnyTwoImageBufferOptionsTest::imageBufferOptions1 const):
* Tools/TestWebKitAPI/Tests/WebCore/cg/GraphicsContextTestsCG.cpp:
(TestWebKitAPI::TEST):
* LayoutTests/fast/canvas/small-canvas-layer-putImageData-expected.html:
* LayoutTests/fast/canvas/small-canvas-layer-putImageData.html:
  Added.

Canonical link: <a href="https://commits.webkit.org/260927@main">https://commits.webkit.org/260927@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a075ecf9187c68e57cb80484845c0227edb2406f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109615 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1103 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118745 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9936 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101886 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115370 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15039 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98274 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97013 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29927 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11473 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31269 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12131 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8201 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50878 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7576 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13871 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on LayoutTests/fast/canvas/small-canvas-layer-putImageData.html:
```diff
@@ -0,0 +1,29 @@
+        data[4 * i + 2] = 0;
+        data[4 * i + 3] = 255;
+    }
+    ctx.putImageData(imageData, 0, 0);
```
I think this test need `waitUntilDone()` and `notifyDone()` since it requires at two rAFs to be served.

kkinnunen-apple replied:
thanks!

shallawa commented on Source/WebCore/platform/graphics/cg/GraphicsContextPlatformPrivateCG.h:
```diff
@@ -32,7 +32,6 @@ namespace WebCore {
 enum class GraphicsContextCGFlag : uint8_t {
     IsLayerCGContext        = 1 << 0,
-    IsAcceleratedCGContext  = 1 << 1
 };
```
I think this enum can be removed and `m_contextFlags` can be just a bool `m_isCALayerContext` or something else.

shallawa commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -1472,15 +1472,9 @@ bool GraphicsContextCG::isCALayerContext() const
 RenderingMode GraphicsContextCG::renderingMode() const
 {
-    return m_data->m_contextFlags.contains(GraphicsContextCGFlag::IsAcceleratedCGContext) ? RenderingMode::Accelerated : RenderingMode::Unaccelerated;
+    return CGContextGetType(m_data->m_cgContext.get()) == kCGContextTypeIOSurface ? RenderingMode::Accelerated : RenderingMode::Unaccelerated;
```
You said in the change log that "CoreAnimation might give a bitmap-backed CGContext for the drawing context, even though WebKit requested an accelerated context."

Did you check that `CGContextGetType()` will not return `kCGContextTypeIOSurface` for the context even if it was created by calling `CGIOSurfaceContextCreate`?

In other words, is it sufficient always to test the return value of `CGContextGetType()` to check whether the context is accelerated or not?

kkinnunen-apple replied:
>You said in the change log that "CoreAnimation might give a bitmap-backed CGContext for the drawing context, even though WebKit requested an accelerated context."

Yes, CoreAnimation does conceptually
```
if (size == small)
  return CGBitmapContextCreate();
return CGIOSurfaceContextCreate();
```

> Did you check that CGContextGetType() will not return kCGContextTypeIOSurface for the context even if it was created by calling CGIOSurfaceContextCreate?

CGContextGetType() is the "class type" accessor of the instance and it's used in many places to disambiguate the context type reliably. As far as I can read the code, yes, it's reliable.

kkinnunen-apple replied:
I see what you meant there. Yeah, it appears CALayer marks its IOSurface context as Unknown. Adjusted the patch.

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/ImageBufferTests.cpp:
```diff
@@ -44,7 +45,7 @@ static ::testing::AssertionResult imageBufferPixelIs(Color expected, ImageBuffer
     if (got != expected) {
         // Use this to debug the contents in the browser.
-        // WTFLogAlways("%s", imageBuffer.toDataURL("image/png"_s).latin1().data());
+        WTFLogAlways("%s", imageBuffer.toDataURL("image/png"_s).latin1().data());
```
Did you uncomment this statement unintentionally?

kkinnunen-apple replied:
Yes, reverted, thansk!

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/cg/GraphicsContextTestsCG.cpp:
```diff
@@ -74,6 +75,29 @@ TEST(GraphicsContextTests, DrawNativeImageDoesNotLeakCompositeOperator)
+    auto cgContext = adoptCF(CGIOSurfaceContextCreate(surface->surface(), size.width(), size.height(), bitsPerComponent, bitsPerPixel, srgb.platformColorSpace(), bitmapInfo));
+    GraphicsContextCG context(cgContext.get());
+    EXPECT_EQ(context.renderingMode(), RenderingMode::Accelerated);
+}
```
It is strange that this test creates a context with size = { 3, 3 } and `context.renderingMode()` still returns `Accelerated`. Is there a way to get and test the case when this context is unaccelerated?

kkinnunen-apple replied:
> Is there a way to get and test the case when this context is unaccelerated?

Yes, that's the case above.

> It is strange that this test creates a context with size = { 3, 3 } and context.renderingMode() still returns Accelerated. 

If the context is IOSurface context, it's accelerated.
If the context is CGBitmap context, it's unaccelerated.
Size has nothing to do with it.

kkinnunen-apple replied:
Added a corresponding test for the small and large CALayer

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/ImageBufferTests.cpp:
```diff
@@ -336,4 +393,11 @@ INSTANTIATE_TEST_SUITE_P(ImageBufferTests,
+INSTANTIATE_TEST_SUITE_P(ImageBufferTests,
+    AnyTwoImageBufferOptionsTest,
+    testing::Combine(
+        testing::Values(TestImageBufferOptions::NoOptions, TestImageBufferOptions::Accelerated),
+        testing::Values(TestImageBufferOptions::NoOptions, TestImageBufferOptions::Accelerated)),
```
Did you mean to pass the same `TestImageBufferOptions::Accelerated` for the two tests?

kkinnunen-apple replied:
Yes, that's the intention. Combine is a cartesian product.
`Combine({a,b}, {a,b})` means `{a,b} x {a,b}`, e.g it generates list `{aa, ab, ba, bb}`. 
So this is what we want: test all the possible combinations of source and destination.

kkinnunen-apple committed:
MotionMark Images subtest rendering is incorrect on ASi (no dissolve effect)
https://bugs.webkit.org/show_bug.cgi?id=252482
rdar://problem/105606592

Reviewed by Simon Fraser.

Fix rendering of accelerated canvas elements into small layer tiles
after putImageData.

ImageBufferIOSurfaceBackend::putPixelBuffer would clean the CGImage
cache, if there's possibility that the cache exists.

ImageBufferIOSurfaceBackend::copyNativeImageForDrawing() does not
currently mark the cache as being active. Instead, it relies on
after-the-fact cleanup via
ImageBufferIOSurfaceBackend::finalizeDrawIntoContext().

ImageBufferIOSurfaceBackend::finalizeDrawIntoContext() would only
cleanup if the draw happened on an non-accelerated context.

Accelerated context status would be determined by whether WebKit
requested the layer to be accelerated.

In case the layer is small, CoreAnimation might give a bitmap-backed
CGContext for the drawing context, even though WebKit
requested an accelerated context. WebKit would
mark the context as accelerated, even when it was not.

This would lead into scenario where the CGImage cache would not be
cleaned after each draw, and hence putImageData would not be visible.

Fix by removing the GraphicsContextCG::setIsAcceleratedContext().
Instead, query the actual state from the CGContext.

Adds a test in ImageBufferTests that tests the code paths. This test
would not have caught the issue, since the bug as incorrect use of
setIsAcceleratedContext() rooted from PlatformCALayer::drawLayerContents.
Nevertheless, the test would have given a bit more confidence that the
complicated invalidation logic was correct in the specific case.

Adds a trivial test for the acceleration status query based on the
CGContext.

* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::setIsAcceleratedContext): Deleted.
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayer::drawLayerContents):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::renderingMode const):
(WebCore::GraphicsContextCG::setIsAcceleratedContext): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextPlatformPrivateCG.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::ensureGraphicsContext):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/mac/WebLayer.mm:
(-[WebSimpleLayer drawInContext:]):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _drawRect:contentsOnly:]):
* Tools/TestWebKitAPI/Tests/WebCore/ImageBufferTests.cpp:
(TestWebKitAPI::imageBufferPixelIs):
(TestWebKitAPI::hasTestPattern):
(TestWebKitAPI::drawTestPattern):
(TestWebKitAPI::createPixelBufferTestPattern):
(TestWebKitAPI::toImageBufferOptions):
(TestWebKitAPI::PreserveResolutionOperationTest::imageBufferOptions const):
(TestWebKitAPI::TEST_P):
(TestWebKitAPI::AnyScaleTest::imageBufferOptions const):
(TestWebKitAPI::AnyTwoImageBufferOptionsTest::imageBufferOptions0 const):
(TestWebKitAPI::AnyTwoImageBufferOptionsTest::imageBufferOptions1 const):
* Tools/TestWebKitAPI/Tests/WebCore/cg/GraphicsContextTestsCG.cpp:
(TestWebKitAPI::TEST):
* LayoutTests/fast/canvas/small-canvas-layer-putImageData-expected.html:
* LayoutTests/fast/canvas/small-canvas-layer-putImageData.html:
  Added.

Canonical link: https://commits.webkit.org/260927@main
